feat(blueprints): automatic side substitution

keithharvey opened the pull request:
Closes #4816 

## What's It Do?
Blueprints are automatically converted to the player’s faction before placement


* New Logic Integration: Added blueprint_substitution/logic.lua and integrated it into `api_blueprint.lua` (runtime blueprints) and `gui_pregame_build.lua` (pregame blueprints)
* Blueprint Category Definition: These group unit ids by "blueprint category", e.g. "SOLAR" or "ADVANCED_SOLAR" etc.
* 
* Reports: Introduced optional dev reports for working with units in bulk

## Testing

* [X] create a blueprint in one faction (`/nocost` helps a lot)
* [X] re-enter the game as another faction
* [X] use same blueprint and verify the mappings all worked

## Screenshots
### T1
#### Core/OG
![Screenshot 2025-04-22 193821](https://github.com/user-attachments/assets/3aafe96e-d16c-4614-9f72-c66a35c525ad)

#### Arm
![Screenshot 2025-04-22 201416](https://github.com/user-attachments/assets/f2ad06c9-c788-4dea-9ad0-a9073c1c9f67)

#### Legion
![Screenshot 2025-04-22 200422](https://github.com/user-attachments/assets/21771ca1-84b1-4ebb-8d11-f8b0a4c2e40f)

### T2
#### Core/OG
![image](https://github.com/user-attachments/assets/3c3c148c-0070-4f4e-89a8-0dd831de1464)

#### Arm
![image](https://github.com/user-attachments/assets/59fdcc93-8284-454f-bc7f-fa730be8e666)

#### Legion
![Screenshot 2025-04-22 215740](https://github.com/user-attachments/assets/10f7d375-8fe2-475d-aec8-44c9d85ce4d9)

### Modified Echos

#### Failure to Substitute
![image](https://github.com/user-attachments/assets/d6664142-021e-4e71-8c86-74818be3918b)

#### Success
![image](https://github.com/user-attachments/assets/f4e87d77-a954-4b96-8449-908277eda2c0)

## Future Work: Completing Unit Category Mappings
Not all units (especially underwater ones!) have mappings.

### Manual Mapping

What I did was make a blueprint in one faction then fire up that same blueprint in another faction and go through the failures in the logs until I had them mapped.

### Duplicate Categories defined by mistake?

It will warn you in the logs whenever you duplicate a category. We really want to avoid people accidentally ghosting already-defined categories and catch it immediately during development or testing. For example

```lua
    DefCat("UW_METAL_STORAGE", {[SIDES.ARM]="armuwms", [SIDES.CORE]="coruwms", [SIDES.LEGION]="legamstor"})
    DefCat("UW_METAL_STORAGE", {[SIDES.ARM]="danger", [SIDES.CORE]="will", [SIDES.LEGION]="robinson"})
```
results in
```zsh
[2025-04-22 23:34:01.089] [info]  [t=31:50:40.278055][f=-000001] Failed to load: api_blueprint.lua  ([LuaVFS::Include(synced=false)][pcall] file=luaui/Include/blueprint_substitution/logic.lua error=2 ([string "luaui/Include/blueprint_substitution/definitions.lua"]:62: [BlueprintDefs ERROR] Duplicate category key definition attempted: 'UW_METAL_STORAGE') ptop=1 cenv=false vfsmode=rMmeb)
[2025-04-22 23:34:01.089] [warn]  [t=31:50:40.278001][f=-000001] [BlueprintDefs] Error: [BlueprintDefs ERROR] Duplicate category key definition attempted: 'UW_METAL_STORAGE'
```

### Diagnosing Mapping Failure

In game, run `/blueprintmapreport` (a per-unit csv) and `/blueprintcategorylist` (a unit category focused list of mappings)

Check the logs (where you can also find the full reports)
```zsh
tail -f $(ls -t /c/Users/$USERNAME/AppData/Local/Programs/Beyond-All-Reason/data/launcher-logs/spring-launcher-*.log | head -1)
....
[2025-04-22 21:35:47.711] [info]  [t=29:52:26.899087][f=0001149] [BlueprintSubLogic WARNING] 4 units failed substitution
[t=29:52:26.899149][f=0001149] [Blueprint API] Substitution Result: Blueprint conversion from cor to arm attempted. Total: 20, Skipped: 0, Attempted: 20, Success: 16, FailedMap: 4, FailedLookup: 0 (FAIL)
[2025-04-22 21:35:47.711] [warn]  [t=29:52:26.899100][f=0001149] [BlueprintSubLogic] Warning: Substitution FAILED: corbuzz (Calamity) [corbuzz] from cor to arm - Reason: no_mapping
[t=29:52:26.899111][f=0001149] [BlueprintSubLogic] Warning: Substitution FAILED: corafus (Advanced Fusion Reactor) [corafus] from cor to arm - Reason: no_mapping
[t=29:52:26.899119][f=0001149] [BlueprintSubLogic] Warning: Substitution FAILED: cortron (Catalyst) [cortron] from cor to arm - Reason: no_mapping
[t=29:52:26.899127][f=0001149] [BlueprintSubLogic] Warning: Substitution FAILED: cortoast (Persecutor) [cortoast] from cor to arm - Reason: no_mapping
```

## Design Considerations

### Defining Categories on Units Definitions#customParams?

I considered it for a while. Making this purely data-driven in that way makes a lot of sense, especially if we find a need for categories in other areas of the game. However, for now there are substitutions that only make sense in the context of blueprints and having them in one, category-centric place made sense.

keithharvey commented on luaui/Include/blueprint_substitution/logic.lua:
```diff
@@ -0,0 +1,489 @@
+
+    for k in pairs(equivalentUnits) do equivalentUnits[k] = nil end -- Clear before generating
+
+    local mappingCount = 0
```
would love functional constructs here, is that a thing in lua? Map/reduce/filter?

sprunk replied:
check these out
https://github.com/beyond-all-reason/Beyond-All-Reason/blob/1d8555f625ab247632c0c1a86be239b1b6c4afb7/common/tablefunctions.lua#L311-L382

keithharvey replied:
Nice. Ty

keithharvey replied:
That helped a lot. I am now done refactoring if you want to take another look. We were able to pre-munge a "master building list" with our equivalences mapped out for every possible faction permutation. This simplified everything and let us keep it more map-reducey/flat.

keithharvey commented:
Whoops. Honestly mine looks slightly more polished than either but that's a good callout I should've looked around.

keithharvey commented:
Rebased

keithharvey commented on luaui/Include/blueprint_substitution/logic.lua:
```diff
@@ -0,0 +1,459 @@
+                                 targetSide,
+                                 #originalBlueprint.units))
+    end
+
```
$$$$$$$$$$$$$

keithharvey commented:
@sprunk I went ahead and read #3524 and integrated any feedback from there I thought would help here, but please let me know if I missed anything. The only thing I'm really not sure about here is my factoring and whether or not all of this belongs in `luaui/include`, but seems in line with what you were suggesting he do, just might not be in the right place.

keithharvey commented on luaui/Widgets/api_blueprint.lua:
```diff
@@ -1,5 +1,16 @@
+
+local reportFunctions = nil
+
+local activeBlueprint = nil
```
moved this to the top of the file to reflect that it is now "global state", insomuch as it is accessed by cmd_blueprint via a getter.

keithharvey commented on luaui/Include/blueprint_substitution/reports.lua:
```diff
@@ -0,0 +1,182 @@
+-- luaui/Include/blueprint_substitution/reports.lua
+-- Contains reporting functions for api_blueprint.lua
+
```
preserved these simply because we know the mappings are incomplete, and they were useful for programmatic processing.

keithharvey commented on luaui/Include/blueprint_substitution/logic.lua:
```diff
@@ -0,0 +1,369 @@
+    end
+end
+Spring.Log("BlueprintSubLogic", "info", "Finished precomputing Name->ID map.")
+
```
This was useful to "enforce" constants, but is only strict during debug, and even then it just errors louder/does more checks.

sprunk replied:
What's the worst that can happen if you're not strict?

keithharvey replied:
Someone commits a file with multiple entries for the same category, where the first category is mistakenly ghosted and everyone ignores the log warnings/errors during debug.

keithharvey commented on luaui/Include/blueprint_substitution/definitions.lua:
Note: I could imagine a world where 
1) these categories are more complete
2) we discover other uses for them
3) they get moved out of our category-centric `definitions.lua` and onto the unit definitions themselves.

keithharvey commented on luaui/Widgets/api_blueprint.lua:
```diff
@@ -687,13 +697,42 @@ end
 	end
 
-	activeBlueprint = bp
+	local blueprintToProcess = table.copy(bp)
```
Okay, incorporating your points:

**Why `table.copy()`:** A deep copy (`table.copy`) was added in `setActiveBlueprint`. This prevents the substitution logic from modifying the original blueprint data held by `cmd_blueprint`.

**Resulting Change:** The `activeBlueprint` in `api_blueprint` now always holds the processed (copied and potentially substituted) version for display.

**Impact on Placement:** `cmd_blueprint` now calls `api_blueprint.getActiveBlueprint()` before placing, ensuring it uses the correctly substituted blueprint. This avoids using its own potentially stale reference and crucially prevents the modified blueprint from being saved back to the file on shutdown.

keithharvey commented:
@sprunk I went through this again and cleaned up a few minor things that bugged me, and tested again. I cleaned up my commit history and put specific comments where I thought they were useful to defend non-obvious decisions. Also incorporated anything I thought was a good idea from the other two PRs (and added a Co-author footer to the commit). Please let me know if you would like any changes or have any questions.

keithharvey commented on luaui/Include/blueprint_substitution/definitions.lua:
```diff
@@ -0,0 +1,166 @@
+-- luaui/Include/blueprint_substitution/definitions.lua
```
![image](https://github.com/user-attachments/assets/55971f3b-84c9-448e-ba71-a2260cb88269)
Not entirely sold on this factoring/naming. I could see us wanting to call definitions "category_definitions" and logic "blueprint_substitution_logic", for easier finding. This was really just me doing the factoring I'm used to/prefer, which avoids redundancy in filenames. For logic and definitions, I could also see us just embedding these directly in a gadget? Just let me know I didn't spend too much time thinking about this particular problem.

keithharvey commented:
@sprunk I went through this again and cleaned up a few minor things that bugged me, and tested again. I cleaned up my commit history and put specific comments where I thought they were useful to defend non-obvious decisions. Also incorporated anything I thought was a good idea from the other two PRs (and added a Co-author footer to the commit). Please let me know if you would like any changes or have any questions.

sprunk commented on luaui/Include/blueprint_substitution/reports.lua:
```diff
@@ -0,0 +1,182 @@
+end
+
+function M.generateCategoryListReport()
+    Spring.Echo("\n--- Blueprint Category List Start ---")
```
How are people supposed to use this? IIRC the console only shows like 2 lines tops so a big dump is probably not very readable. If they are supposed to read this from infolog then consider building a giant string (perhaps via `table.concat`) and doing a single Echo, because that way the engine won't append its `[t=0.0123456][frame=123123]` garbage before each line which makes it difficult to work with the output.

keithharvey replied:
> How are people supposed to use this?
`/blueprintcategorylist` and `/blueprintmapreport`

Will do.

sprunk replied:
>> How are people supposed to use this?
>
> `/blueprintcategorylist` and `/blueprintmapreport`

I meant more in the sense of, when do they call one of the above (rarely, just to verify things, meaning it doesn't need to be super fancy? or more commonly?) and what do they do with its output (just a human glance, or does it make sense to make it machine parse friendly?).

sprunk commented on luaui/Include/blueprint_substitution/reports.lua:
```diff
@@ -0,0 +1,182 @@
+    local masterBuildingDataMinimal = substitutionLogic.MasterBuildingData
+
+    local reportHeader = table.concat({
+        "Category", "DescName", -- "NameSource", -- Removing NameSource for now
```
If it's not there then just get rid of it.
```suggestion
        "Category", "DescName",
```

sprunk commented on luaui/Include/blueprint_substitution/reports.lua:
```diff
@@ -0,0 +1,182 @@
+local sideComplete = {}
+local uncategorizedUnits = {}
+
+local function escapeCsvField(value)
```
Could eventually go into some sort of `Spring.Utilities`, though not in the scope of this PR.

sprunk commented on luaui/Include/blueprint_substitution/reports.lua:
```diff
@@ -0,0 +1,182 @@
+-- luaui/Include/blueprint_substitution/reports.lua
+-- Contains reporting functions for api_blueprint.lua
+
+local M = {}
```
I expect @WatchTheFort would be unhappy with `M` as a file scope variable name.

keithharvey commented on luaui/Widgets/api_blueprint.lua:
```diff
@@ -687,13 +697,42 @@ end
+				Spring.Echo("[Blueprint Detail] " .. result.summaryMessage)
+			end
+		else
+			Spring.Log("BlueprintAPI", "warning", "processBlueprintSubstitution returned nil for targetSide: " .. tostring(targetSide))
+			Spring.Echo("[Blueprint API] Substitution processing failed for target side: " .. tostring(targetSide))
```
The log I treat as "log" and echo I treat as "communicate with the user". Not sure that's 100% canonical so please correct me if I'm wrong. This was also a mistake since I assume echo goes to the logs.

sprunk replied:
I don't think a "canonical" way exists so echo is as good a choice for user feedback as any other, but echo has a ton of use cases and some overlap with log so it would be nice to convey the meaning via something like
```lua
local FeedbackForUser = Spring.Echo -- could later be improved with fancy graphical UI or something
```

keithharvey replied:
I also extended this pattern to cmd_blueprint for module wide consistency.

sprunk commented on luaui/Include/blueprint_substitution/definitions.lua:
```diff
@@ -0,0 +1,164 @@
+-- ===================================================================
+Spring.Log("BlueprintDefs", LOG.INFO, "Initializing blueprint category definitions...")
+
+-- Define sides directly here since we're in unsynced code and can't access gamedata/sidedata.lua
```
You can actually access `gamedata/sidedata.lua` here (either via `Spring.GetSideData` or `VFS.Include`) but I'm not sure how useful it would be.

keithharvey replied:
Yeah originally I had started to modify that file to have an actual completion/match my own opinion that enums should be inclusive and represent real values, but then I noped out. Right now that seems to be a data structure driving the side selection options. It wasn't that hard maybe one day.

keithharvey replied:
I think just do the refactor? The more I stare at this the more it makes sense just to refactor it?

keithharvey replied:
Do we like Sides here? I was unsure what ubiquitous language was here. It's firmly sides in code but factions in the UI. We could take this opportunity to rename sides to factions if we're touching sidedata?

sprunk commented on luaui/Include/blueprint_substitution/definitions.lua:
```diff
@@ -0,0 +1,164 @@
+BpDefs.SIDES = {
+    ARM = "arm",
+    CORE = "cor",
+    LEGION = "leg"
+}
```
Seems to be missing the UNKNOWN side.

keithharvey replied:
Regarding SIDES vs ["UNKNOWN"]: You're right that functionally, "UNKNOWN" acts like a first-class citizen within the context of uncategorizedUnits. If the goal was purely local consistency within reports.lua, adding it to SIDES might seem logical.

However, I prefer keeping the sentinel value separate from the main SIDES definition from a domain-driven design perspective. Adding "UNKNOWN" there pollutes the core domain enum, which should ideally only represent actual factions. Keeping SIDES clean provides immediate benefits, like the ability to cleanly enumerate over just the valid factions.

Looking ahead, my goal would be to transition towards a more global SIDES enum, and that definition should strictly include only valid sides. Keeping "UNKNOWN" separate now supports that cleaner future representation. So while it creates a slight structural inconsistency here, it maintains the integrity of the SIDES definition itself. Does that make sense?

sprunk replied:
I don't have a strong opinion, maybe @WatchTheFort would like to say something.

WatchTheFort replied:
I don't follow what's being talked about. Is this about supporting more than three hardcoded factions?

keithharvey replied:
I dont think this is a big deal. We ended up solving this by just adding a proper factions enum, which I like quite a bit.

sprunk commented on luaui/Include/blueprint_substitution/logic.lua:
```diff
@@ -0,0 +1,356 @@
+
+        local buildingData = BlueprintSubLogic.MasterBuildingData[unitNameLower]
+
+        if buildingData then
+            record.sourceSide = buildingData.side
+            record.translatedHumanName = buildingData.translatedHumanName
+
+            if record.sourceSide and record.sourceSide ~= targetSide then
+                record.equivalentUnitName = buildingData.equivalents[targetSide]
+
+                if record.equivalentUnitName then
+                    local foundDefID = unitNameToDefIDMap[record.equivalentUnitName]
+                    if foundDefID then
+                        record.status = "success"
+                        record.newUnitDefID = foundDefID
+                    else
+                        record.status = "failed_lookup"
+                    end
+                else
+                    record.status = "failed_no_mapping"
+                end
+            else
+                record.status = "skipped_same_side" 
+            end
+        else
+            record.status = "skipped_not_building"
+        end
```
This giant pyramid looks like it could be a function. All the statuses would be returns.

sprunk commented on luaui/Include/blueprint_substitution/logic.lua:
```diff
@@ -0,0 +1,356 @@
+            record.status = "skipped_not_building"
+        end
+        
+        -- Use LOG.DEBUG directly
+        Spring.Log("BlueprintSubLogic", LOG.DEBUG, string.format("Mapped unit %s: Status=%s", 
```
Comment not precise enough
```suggestion
        -- Use LOG.DEBUG directly and also call string.format with two %s format specifiers
        Spring.Log("BlueprintSubLogic", LOG.DEBUG, string.format("Mapped unit %s: Status=%s", 
```

keithharvey replied:
Is it ok if I just delete it? That suggested change feels ironically mocking but I honestly can't tell. That comment was gemini 2.5 being derpy, verbose, and disobedient.

sprunk commented on luaui/Widgets/api_blueprint.lua:
```diff
@@ -1,5 +1,15 @@
+local BpDefs = VFS.Include("luaui/Include/blueprint_substitution/definitions.lua")
+VFS.Include("luaui/Include/blueprint_substitution/logic.lua")
+
+local isDevModeParam = Spring.GetGameRulesParam('isDevMode')
+local ENABLE_REPORTS = (isDevModeParam == 1 or isDevModeParam == true or isDevModeParam == "1")
```
```suggestion
local ENABLE_REPORTS = Spring.Utilities.IsDevMode()
```

sprunk commented:
The PR seems to assume that a unit can be in at most 1 faction and have at most 1 role. I'm not necessarily suggesting to change it though, just a loose remark about a tradeoff to be aware of.

WatchTheFort commented:
Both this and pre-game build should be sharing one set of logic for cross-faction equivalent building substitution. Can this be refactored so the logic is not duplicated, and so the configuration data lives in a dedicated file?

keithharvey commented:
@WatchTheFort done

keithharvey commented:
@WatchTheFort done. Tested it again. That was fun.

keithharvey commented:
Thank you so much sprung!

keithharvey commented:
Should I resolve these comments? They were mostly meant as flags for review..

WatchTheFort commented:
> Should I resolve these comments? They were mostly meant as flags for review..

Comments should be resolved:

- By you, when you have fixed the issue the reviewer raised
- By the reviewer, when you give a reason the issue they raised is not actually a problem
